File: headers_list/src/main/java/HeadersList/HeadersList.java
package HeadersList;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.function.Consumer;

//ъуъ
public class HeadersList<T extends Comparable<T>> implements Iterable<HeadersList.SerializableNode<T>> {

    //башку в узел завязал(пара заголовок + список, надо для линковки)
    private static class HeaderNode<T> {
        T header;
        HeaderNode<T> next;
        List<T> associatedList;

        HeaderNode(T header, List<T> list) {
            this.header = header;
            this.associatedList = list;
            this.next = null;
        }
    }

    public record SerializableNode<T>(T header, List<T> associatedList) {}

    private HeaderNode<T> root;
    private int size;

    public HeadersList() {}

    @Override
    public Iterator<SerializableNode<T>> iterator() {
        return new HeadersListIterator();
    }

    private class HeadersListIterator implements Iterator<SerializableNode<T>> {
        private HeaderNode<T> current = root;

        @Override
        public boolean hasNext() {
            return current != null;
        }

        @Override
        public SerializableNode<T> next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            SerializableNode<T> data = new SerializableNode<>(current.header, current.associatedList);
            current = current.next;
            return data;
        }
    }

    //в конец добавляем списачек и загаловочек
    public void add(T header, List<T> list) {
        if (root == null) {
            root = new HeaderNode<>(header, list);
        } else {
            HeaderNode<T> currNode = root;
            while (currNode.next != null) {
                currNode = currNode.next;
            }
            currNode.next = new HeaderNode<>(header, list);
        }
        size++;
    }

    //тотальнейшая вставка списка, его башки ослиной по индексу
    public void add(int index, T header, List<T> list) {
        if (index < 0 || index > size) throw new IndexOutOfBoundsException("ты чё твориш...");
        HeaderNode<T> newNode = new HeaderNode<>(header, list);
        if (index == 0) {
            newNode.next = root;
            root = newNode;
        } else {
            HeaderNode<T> currNode = root;
            for (int i = 0; i < index - 1; i++) {
                currNode = currNode.next;
            }
            newNode.next = currNode.next;
            currNode.next = newNode;
        }
        size++;
    }

    //сортиров очка + балансиров очка
    public void sort() {
        if (size >= 2) {
            HeaderNode<T> last = root;
            while (last.next != null) {
                last = last.next;
            }
            quickSort(root, last);
        }
    }

    private void quickSort(HeaderNode<T> start, HeaderNode<T> end) {
        if (start == null || end == null || start == end || start == end.next) {
            return;
        }

        HeaderNode<T> pivot = partition(start, end);

        if (pivot != start) {
            HeaderNode<T> prevPivot = start;
            while (prevPivot.next != pivot) {
                prevPivot = prevPivot.next;
            }
            quickSort(start, prevPivot);
        }

        if (pivot != end && pivot.next != null) {
            quickSort(pivot.next, end);
        }
    }

    private HeaderNode<T> partition(HeaderNode<T> start, HeaderNode<T> end) {
        T pivotValue = end.header;
        HeaderNode<T> i = start;

        HeaderNode<T> j = start;
        while (j != end) {
            if (j.header.compareTo(pivotValue) < 0) {
                swapNodes(i, j);
                i = i.next;
            }
            j = j.next;
        }

        swapNodes(i, end);
        return i;
    }

    private void swapNodes(HeaderNode<T> node1, HeaderNode<T> node2) {
        if (node1 != node2) {
            T tempHeader =  node1.header;
            List<T> tempList = node1.associatedList;

            node1.header = node2.header;
            node1.associatedList = node2.associatedList;

            node2.header = tempHeader;
            node2.associatedList = tempList;
        }
    }

    public void balance() {
        if (root == null || root.next == null) {
            return;
        }

        List<T> allItems = new ArrayList<>();
        HeaderNode<T> current = root;
        while (current != null) {
            allItems.addAll(current.associatedList);
            current = current.next;
        }

        current = root;
        while (current != null) {
            current.associatedList.clear();
            current = current.next;
        }

        int totalItems = allItems.size();
        int numLists = this.size;
        int baseSize = totalItems / numLists;
        int remainder = totalItems % numLists;

        int currentItemIndex = 0;
        int nodeIndex = 0;
        current = root;

        while (current != null) {
            int sublistSize = baseSize + (nodeIndex < remainder ? 1 : 0);

            for (int j = 0; j < sublistSize; j++) {
                if (currentItemIndex < totalItems) {
                    current.associatedList.add(allItems.get(currentItemIndex++));
                }
            }

            nodeIndex++;
            current = current.next;
        }
    }

    //сортир... Нажмите для продолжения...
    public void addSorted(T header, List<T> list) {
        HeaderNode<T> newNode = new HeaderNode<>(header, list);

        if (root == null) {
            root = newNode;
            size++;
            return;
        }

        int comparison = header.compareTo(root.header);
        if (comparison < 0) {
            newNode.next = root;
            root = newNode;
            size++;
            return;
        }

        if (comparison == 0) {
            root.associatedList.addAll(list);
            return;
        }

        HeaderNode<T> current = root;

        while (current.next != null) {
            comparison = header.compareTo(current.next.header);

            if (comparison == 0) {
                current.next.associatedList.addAll(list);
                return;
            }

            if (comparison < 0) {
                newNode.next = current.next;
                current.next = newNode;
                size++;
                return;
            }
            current = current.next;
        }
        current.next = newNode;
        size++;
    }

    //бошку открути
    public T getHeader(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("ты чё делаешь? мужик, успокойся");
        }
        HeaderNode<T> curr = root;
        for (int i = 0; i < index; i++) {
            curr = curr.next;
        }
        return curr.header;
    }

    //вы списков продаете?
    //нет тока показываю
    //красивое...
    public List<T> getList(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("ты чё делаешь? мужик, успокойся");
        }
        HeaderNode<T> curr = root;
        for (int i = 0; i < index; i++) {
            curr = curr.next;
        }
        return curr.associatedList;    }

    //списачик палучаим дада
    public List<T> getListByHeader(T header) {
        HeaderNode<T> curr = root;
        while (curr != null) {
            if (curr.header.equals(header)) {
                return curr.associatedList;
            }
            curr = curr.next;
        }
        return null;
    }

    //минус элемент
    public void remove(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("да харош уже нет такого индекса");
        }

        if (index == 0) {
            root = root.next;
            size--;
            return;
        }
        HeaderNode<T> prev = root;
        for (int i = 0; i < index - 1; i++) {
            prev = prev.next;
        }
        HeaderNode<T> removingNode = prev.next;
        prev.next = removingNode.next;
        size--;
    }

    //да это на морозе уменьшился чесно
    public int size() {
        return size;
    }

    //вешаем итератор
    public void forEachHeader(Consumer<T> action) {
        HeaderNode<T> curr = root;
        while (curr != null) {
            action.accept(curr.header);
            curr = curr.next;
        }
    }

    //парсуем, я сказала парсуем!
    public void saveToFile(File file) throws IOException {
        List<SerializableNode<T>> dataToSave = new ArrayList<>();

        HeaderNode<T> current = root;
        while (current != null) {
            dataToSave.add(new SerializableNode<>(current.header, current.associatedList));
            current = current.next;
        }

        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        try (FileWriter writer = new FileWriter(file)) {
            gson.toJson(dataToSave, writer);
        }
    }

    public static <T extends Comparable<T>> HeadersList<T> loadFromFile(File file, Type type) throws IOException {
        Gson gson = new Gson();
        HeadersList<T> resultList = new HeadersList<T>();
        try (FileReader reader = new FileReader(file)) {
            Type listType = TypeToken.getParameterized(List.class, type).getType();
            List<SerializableNode<T>> loadedData = gson.fromJson(reader, listType);

            if (loadedData != null) {
                for (SerializableNode<T> nodeData : loadedData) {
                    resultList.add(nodeData.header(), nodeData.associatedList());
                }
            }
        }
        return resultList;
    }

    @Override
    public String toString() {
        if (root == null) {
            return "HeadersList{[]}";
        }

        StringBuilder sb = new StringBuilder();
        sb.append("HeadersList{[");
        HeaderNode<T> current = root;
        while (current != null) {
            sb.append(current.header.toString())
                    .append(": ")
                    .append(current.associatedList.toString());
            if (current.next != null) {
                sb.append("], [");
            }
            current = current.next;
        }
        sb.append("]}");
        return sb.toString();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }

        HeadersList<?> other = (HeadersList<?>) obj;

        if (this.size != other.size) {
            return false;
        }

        HeaderNode<?> currentThis = this.root;
        HeaderNode<?> currentOther = other.root;

        while (currentThis != null) {
            if (!Objects.equals(currentThis.header, currentOther.header) ||
                    !Objects.equals(currentThis.associatedList, currentOther.associatedList)) {
                return false;
            }
            currentThis = currentThis.next;
            currentOther = currentOther.next;
        }
        return true;
    }
    @Override
    public int hashCode() {
        int result = 1;
        HeaderNode<T> current = root;

        while (current != null) {
            int headerHash = Objects.hashCode(current.header);
            int listHash = Objects.hashCode(current.associatedList);

            result = 31 * result + headerHash;
            result = 31 * result + listHash;

            current = current.next;
        }
        return result;
    }
}

File: headers_list/src/main/kotlin/HeadersList/HeadersList.kt
package HeadersList

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.lang.reflect.Type
import java.util.*

//ъуъ
class HeadersList<T : Comparable<T>> : Iterable<HeadersList.SerializableNode<T>> {

    //башку в узел завязал(пара заголовок + список, надо для линковки)
    private class HeaderNode<T>(
        var header: T,
        var associatedList: MutableList<T>,
        var next: HeaderNode<T>? = null
    )

    data class SerializableNode<T>(val header: T, val associatedList: List<T>)

    private var root: HeaderNode<T>? = null
    private var size: Int = 0

    override fun iterator(): Iterator<SerializableNode<T>> {
        return HeadersListIterator()
    }

    private inner class HeadersListIterator : Iterator<SerializableNode<T>> {
        private var current: HeaderNode<T>? = root

        override fun hasNext(): Boolean {
            return current != null
        }

        override fun next(): SerializableNode<T> {
            if (!hasNext()) {
                throw NoSuchElementException()
            }
            val data = SerializableNode(current!!.header, current!!.associatedList)
            current = current!!.next
            return data
        }
    }

    //в конец добавляем списачек и загаловочек
    fun add(header: T, list: MutableList<T>) {
        if (root == null) {
            root = HeaderNode(header, list)
        } else {
            var currNode = root
            while (currNode!!.next != null) {
                currNode = currNode.next
            }
            currNode.next = HeaderNode(header, list)
        }
        size++
    }

    //тотальнейшая вставка списка, его башки ослиной по индексу
    fun add(index: Int, header: T, list: MutableList<T>) {
        if (index < 0 || index > size) throw IndexOutOfBoundsException("ты чё твориш...")
        val newNode = HeaderNode(header, list)
        if (index == 0) {
            newNode.next = root
            root = newNode
        } else {
            var currNode = root
            for (i in 0 until index - 1) {
                currNode = currNode!!.next
            }
            newNode.next = currNode!!.next
            currNode.next = newNode
        }
        size++
    }

    //сортиров очка + балансиров очка
    fun sort() {
        if (size >= 2) {
            var last = root
            while (last!!.next != null) {
                last = last.next
            }
            quickSort(root, last)
        }
    }

    private fun quickSort(start: HeaderNode<T>?, end: HeaderNode<T>?) {
        if (start == null || end == null || start == end || start == end.next) {
            return
        }

        val pivot = partition(start, end)

        if (pivot != start) {
            var prevPivot = start
            while (prevPivot!!.next != pivot) {
                prevPivot = prevPivot.next
            }
            quickSort(start, prevPivot)
        }

        if (pivot != end && pivot.next != null) {
            quickSort(pivot.next, end)
        }
    }

    private fun partition(start: HeaderNode<T>, end: HeaderNode<T>): HeaderNode<T> {
        val pivotValue = end.header
        var i: HeaderNode<T>? = start

        var j: HeaderNode<T>? = start
        while (j != end) {
            if (j!!.header < pivotValue) {
                swapNodes(i!!, j)
                i = i.next
            }
            j = j.next
        }

        swapNodes(i!!, end)
        return i
    }

    private fun swapNodes(node1: HeaderNode<T>, node2: HeaderNode<T>) {
        if (node1 != node2) {
            val tempHeader = node1.header
            val tempList = node1.associatedList

            node1.header = node2.header
            node1.associatedList = node2.associatedList

            node2.header = tempHeader
            node2.associatedList = tempList
        }
    }

    fun balance() {
        if (root == null || root!!.next == null) {
            return
        }

        val allItems = ArrayList<T>()
        var current = root
        while (current != null) {
            allItems.addAll(current.associatedList)
            current = current.next
        }

        current = root
        while (current != null) {
            current.associatedList.clear()
            current = current.next
        }

        val totalItems = allItems.size
        val numLists = this.size
        val baseSize = totalItems / numLists
        val remainder = totalItems % numLists

        var currentItemIndex = 0
        var nodeIndex = 0
        current = root

        while (current != null) {
            val sublistSize = baseSize + if (nodeIndex < remainder) 1 else 0

            for (j in 0 until sublistSize) {
                if (currentItemIndex < totalItems) {
                    current.associatedList.add(allItems[currentItemIndex++])
                }
            }

            nodeIndex++
            current = current.next
        }
    }

    //сортир... Нажмите для продолжения...
    fun addSorted(header: T, list: MutableList<T>) {
        val newNode = HeaderNode(header, list)

        if (root == null) {
            root = newNode
            size++
            return
        }

        var comparison = header.compareTo(root!!.header)
        if (comparison < 0) {
            newNode.next = root
            root = newNode
            size++
            return
        }

        if (comparison == 0) {
            root!!.associatedList.addAll(list)
            return
        }

        var current = root

        while (current!!.next != null) {
            comparison = header.compareTo(current.next!!.header)

            if (comparison == 0) {
                current.next!!.associatedList.addAll(list)
                return
            }

            if (comparison < 0) {
                newNode.next = current.next
                current.next = newNode
                size++
                return
            }
            current = current.next
        }
        current.next = newNode
        size++
    }

    //бошку открути
    fun getHeader(index: Int): T {
        if (index !in 0..<size) {
            throw IndexOutOfBoundsException("ты чё делаешь? мужик, успокойся")
        }
        var curr = root
        for (i in 0 until index) {
            curr = curr!!.next
        }
        return curr!!.header
    }

    //вы списков продаете?
    //нет тока показываю
    //красивое...
    fun getList(index: Int): MutableList<T> {
        if (index !in 0..<size) {
            throw IndexOutOfBoundsException("ты чё делаешь? мужик, успокойся")
        }
        var curr = root
        for (i in 0 until index) {
            curr = curr!!.next
        }
        return curr!!.associatedList
    }

    //списачик палучаим дада
    fun getListByHeader(header: T): MutableList<T>? {
        var curr = root
        while (curr != null) {
            if (curr.header == header) {
                return curr.associatedList
            }
            curr = curr.next
        }
        return null
    }

    //минус элемент
    fun remove(index: Int) {
        if (index !in 0..<size) {
            throw IndexOutOfBoundsException("да харош уже нет такого индекса")
        }

        if (index == 0) {
            root = root!!.next
            size--
            return
        }
        var prev = root
        for (i in 0 until index - 1) {
            prev = prev!!.next
        }
        val removingNode = prev!!.next
        prev.next = removingNode!!.next
        size--
    }

    //да это на морозе уменьшился чесно
    fun size(): Int {
        return size
    }

    //вешаем итератор
    fun forEachHeader(action: (T) -> Unit) {
        var curr = root
        while (curr != null) {
            action(curr.header)
            curr = curr.next
        }
    }

    //парсуем, я сказала парсуем!
    @Throws(IOException::class)
    fun saveToFile(file: File) {
        val dataToSave = ArrayList<SerializableNode<T>>()

        var current = root
        while (current != null) {
            dataToSave.add(SerializableNode(current.header, current.associatedList))
            current = current.next
        }

        val gson = GsonBuilder().setPrettyPrinting().create()
        FileWriter(file).use { writer ->
            gson.toJson(dataToSave, writer)
        }
    }

    companion object {
        @JvmStatic
        @Throws(IOException::class)
        fun <T : Comparable<T>> loadFromFile(file: File, type: Type): HeadersList<T> {
            val gson = Gson()
            val resultList = HeadersList<T>()
            FileReader(file).use { reader ->
                val listType = TypeToken.getParameterized(List::class.java, type).type
                val loadedData: List<SerializableNode<T>>? = gson.fromJson(reader, listType)

                if (loadedData != null) {
                    for (nodeData in loadedData) {
                        resultList.add(nodeData.header, nodeData.associatedList.toMutableList())
                    }
                }
            }
            return resultList
        }
    }

    override fun toString(): String {
        if (root == null) {
            return "HeadersList{[]}"
        }

        val sb = StringBuilder()
        sb.append("HeadersList{[")
        var current = root
        while (current != null) {
            sb.append(current.header.toString())
                .append(": ")
                .append(current.associatedList.toString())
            if (current.next != null) {
                sb.append("], [")
            }
            current = current.next
        }
        sb.append("]}")
        return sb.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }

        val otherList = other as HeadersList<*>

        if (this.size != otherList.size) {
            return false
        }

        var currentThis = this.root
        var currentOther = otherList.root

        while (currentThis != null) {
            if (currentThis.header != currentOther!!.header ||
                currentThis.associatedList != currentOther.associatedList
            ) {
                return false
            }
            currentThis = currentThis.next
            currentOther = currentOther.next
        }
        return true
    }

    override fun hashCode(): Int {
        var result = 1
        var current = root

        while (current != null) {
            val headerHash = Objects.hashCode(current.header)
            val listHash = Objects.hashCode(current.associatedList)

            result = 31 * result + headerHash
            result = 31 * result + listHash

            current = current.next
        }
        return result
    }
}
